File: Ticket.cpp
#include "Ticket.hpp"
#include "Trip.hpp"
#include "User.hpp"

using namespace InfSystBusStation;

// Конструктор
Ticket::Ticket(int placeNumber, Trip^ trip, User^ passenger, TicketType type)
    : placeNumber(placeNumber),
    tripData(trip),
    passengerData(passenger),
    type(type),
    ticketAvail(true),
    ticketStatus("Забронирован"),
    finalPrice(0.0)
{
    if (placeNumber <= 0) {
        throw gcnew System::ArgumentException("Номер места должен быть положительным!");
    }

    if (passenger == nullptr) {
        throw gcnew System::ArgumentNullException("Пассажир не может быть null!");
    }

    if (trip == nullptr) {
        throw gcnew System::ArgumentNullException("Поездка не может быть null!");
    }

    CalculateFinalPrice();
    System::Console::WriteLine("[Ticket] Создан билет на место №{0} для {1}",
        placeNumber, passenger->GetFullName());
}

// Расчет конечной цены
void Ticket::CalculateFinalPrice() {
    if (tripData == nullptr) {
        finalPrice = 0.0;
        return;
    }

    double basePrice = tripData->GetPrice();
    finalPrice = CalculatePrice(basePrice, type);

    // Если есть пользователь, применить его скидку
    if (passengerData != nullptr) {
        double discount = passengerData->CalculateDiscount();
        if (discount > 0) {
            double originalPrice = finalPrice;
            finalPrice = finalPrice * (1.0 - discount / 100.0);
            System::Console::WriteLine("[Ticket] Применена скидка {0}%: {1:F2} руб. → {2:F2} руб.",
                discount, originalPrice, finalPrice);
        }
    }
}

// Вывод информации о билете
void Ticket::PrintTicketInfo() {
    System::Console::WriteLine("=== ИНФОРМАЦИЯ О БИЛЕТЕ ===");
    System::Console::WriteLine("Номер места: {0}", placeNumber);
    System::Console::WriteLine("Тип билета: {0}", TicketTypeName);

    if (tripData != nullptr) {
        System::Console::WriteLine("Маршрут: {0}", tripData->GetRoute());

        // Используем существующие методы Trip для получения даты и времени
        System::Console::WriteLine("Дата: {0:dd.MM.yyyy}", tripData->GetTripDate());
        System::Console::WriteLine("Время: {0}", tripData->GetTripTime());

        System::Console::WriteLine("Цена поездки: {0:F2} руб.", tripData->GetPrice());
    }

    if (passengerData != nullptr) {
        System::Console::WriteLine("Пассажир: {0}", passengerData->GetFullName());
        System::Console::WriteLine("Email: {0}", passengerData->GetEmail());
        System::Console::WriteLine("Телефон: {0}", passengerData->PhoneNumber);

        // Если есть паспортные данные
        System::String^ passportSeries = passengerData->GetPassportSeries();
        System::String^ passportNumber = passengerData->GetPassportNumber();
        if (!System::String::IsNullOrEmpty(passportSeries) &&
            !System::String::IsNullOrEmpty(passportNumber)) {
            System::Console::WriteLine("Паспорт: {0} {1}", passportSeries, passportNumber);
        }
    }

    System::Console::WriteLine("Базовая цена: {0:F2} руб.",
        tripData != nullptr ? tripData->GetPrice() : 0.0);
    System::Console::WriteLine("Конечная цена: {0:F2} руб.", finalPrice);
    System::Console::WriteLine("Статус: {0}", ticketStatus);
    System::Console::WriteLine("==========================");
}

File: Ticket.java
public class Ticket {
    public enum TicketType {
        ADULT,      
        CHILD,      
        LUGGAGE     
    }

    private int placeNumber;
    private Trip tripData;
    private Passenger passengerData;
    private boolean ticketAvailable;
    private String ticketStatus;
    private double finalPrice;
    private TicketType type;

    public Ticket(int placeNumber, Trip trip, Passenger passenger, TicketType type) {
        this.placeNumber = placeNumber;
        this.tripData = trip;
        this.passengerData = passenger;
        this.type = type;
        this.ticketAvailable = true;
        this.ticketStatus = "Booked";
        calculateFinalPrice();
    }

    public void calculateFinalPrice() {
        double basePrice = tripData.getPrice();
        double discountMultiplier = 1.0;

        switch (type) {
            case ADULT:
                discountMultiplier = 1.0;  
                break;
            case CHILD:
                discountMultiplier = 0.5;  
                break;
            case LUGGAGE:
                discountMultiplier = 0.3;  
                break;
        }

        finalPrice = basePrice * discountMultiplier;
    }

    public String getTicketTypeName() {
        switch (type) {
            case ADULT:
                return "Adult";
            case CHILD:
                return "Child";
            case LUGGAGE:
                return "Luggage";
            default:
                return "Unknown";
        }
    }

    public void printTicketInfo() {
        System.out.println("=== Ticket Information ===");
        System.out.println("Seat: " + placeNumber);
        System.out.println("Type: " + getTicketTypeName());
        System.out.println("Route: " + tripData.getRoute());
        System.out.println("Price: " + finalPrice + " rub.");
        System.out.println("Passenger: " + passengerData.getFullName());
        System.out.println("Ticket status: " + ticketStatus);

        // Информация по автобусу и водителю
        System.out.println(tripData.getBus() != null ? 
            "Bus: " + tripData.getBus().getBrand() + " [" + tripData.getBus().getCode() + "]" : 
            "Bus: not assigned");

        System.out.println(tripData.getDriver() != null ? 
            "Driver: " + tripData.getDriver().getFullName() : 
            "Driver: not assigned");

        System.out.println("==========================");
    }

    public void ticketPaid() {
        ticketStatus = "Paid";
        System.out.println("Ticket paid");
    }

    // Геттеры
    public int getPlaceNumber() {
        return placeNumber;
    }

    public Trip getTrip() {
        return tripData;
    }

    public Passenger getPassenger() {
        return passengerData;
    }

    public TicketType getTicketType() {
        return type;
    }

    public boolean isAvailable() {
        return ticketAvailable;
    }

    public String getStatus() {
        return ticketStatus;
    }

    public double getFinalPrice() {
        return finalPrice;
    }

    public double getPrice() {
        return finalPrice;
    }

    // Сеттеры
    public void setPlaceNumber(int placeNumber) {
        this.placeNumber = placeNumber;
    }

    public void setTripData(Trip tripData) {
        this.tripData = tripData;
    }

    public void setPassengerData(Passenger passengerData) {
        this.passengerData = passengerData;
    }

    public void setTicketAvailable(boolean ticketAvailable) {
        this.ticketAvailable = ticketAvailable;
    }

    public void setTicketStatus(String ticketStatus) {
        this.ticketStatus = ticketStatus;
    }

    public void setType(TicketType type) {
        this.type = type;
        calculateFinalPrice(); 
    }
}